### PR TITLE
Salesforce integration

### PR DESCRIPTION
**Pull request by justBanks**
Latest work in progress

### CODE REVIEW

**Comment by justBanks** [comment]
I have some things to discuss, as well as some super secret settings to share

**Review by marlonkeating — CHANGES_REQUESTED** [review]

File: salesforce/campaign.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+''' Project model maps to Salesforce Campaign object '''
+client = SalesforceClient()
+
+
+def run(request):
+    response = SalesforceClient().send(request)
+
+
+def save(project: object):
+    stage_tags = list(project.project_stage.all().values())
```
Comment (marlonkeating) [inline salesforce/campaign.py:14]: In order to retrieve a Tag's display name from its slug, first retrieve the Tag object with
https://github.com/DemocracyLab/CivicTechExchange/blob/7a6588ae5417b655e58a448ff023b3523d418345/common/models/tags.py#L23

Once you have the Tag object, you can access the Tag's display name with
https://github.com/DemocracyLab/CivicTechExchange/blob/7a6588ae5417b655e58a448ff023b3523d418345/common/models/tags.py#L7

**Review by marlonkeating — CHANGES_REQUESTED** [review]

**Review by marlonkeating — CHANGES_REQUESTED** [review]

File: salesforce/client.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    def send(self, req):
+        if settings.SALESFORCE_JWT:
+            """ The PreparedRequest has settings merged from the Request instance and those of the Session """
+            prepped_request = self.__session.prepare_request(req)
+            response = self.__session.send(prepped_request)
+            if response.status_code == requests.codes.unauthorized:
+                auth = self.authorize()
+                if auth.status_code == requests.codes.ok:
+                    prepped_request.headers['Authorization'] = SalesforceClient.bearer_token
+                    response = self.__session.send(prepped_request)
```
[the remark this answers is not in the document]
Comment (justBanks) [reply salesforce/client.py:40]: We're about to send a prepped_request so it needs the auth header to be set.  The response just came back unauthorized, so authorize() was called in the next step; that is where the client's bearer_token is set, and hopefully that will persist in the static class variable.

Don't want to hit that unauthorized branch more than once, of course. The auth header is also set in initialize_session, where I expect subsequent instances to pull the token from the client's persisted value.

File: salesforce/campaign.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    issue_area_tags = list(project.project_issue_area.all().values())
+    data = {
+                "ownerid": client.owner_id,
+                "Project_Owner__r":
+                {
+                    "platform_id__c": project.project_creator.id
+                },
+                "recordtypeid": "01246000000uOeRAAU",
+                "name": project.project_name,
+                "type": "Informal (No Legal Entity Established)",
```
[the remark this answers is not in the document]
Comment (justBanks) [reply salesforce/campaign.py:27]: Right you are... I must have been avoiding dealing with tags at the time

File: docker-compose.yml [hunk 4]
```diff
@@ -7,8 +7,6 @@ services:
       - ./.env
     ports:
       - "5432:5432"
-    env_file:
```
[the remark this answers is not in the document]
Comment (justBanks) [reply docker-compose.yml:10]: It's not a duplicate? I had things running fine without it

File: salesforce/campaign.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justBanks) [reply salesforce/campaign.py]: pushed the fix for this

**Review by marlonkeating — APPROVED** [review]

**Pull request merged**